The job **`aarch64-gnu-llvm-21-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/24486182842/job/71561587094) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/71561587094) [(plain)](https://github.com/rust-lang/rust/commit/c3021c37653d0d9bcd71eaba85ee88ff103f8ccc/checks/71561587094/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Executing "/scripts/stage_2_test_set1.sh"
+ /scripts/stage_2_test_set1.sh
PR_CI_JOB set; skipping tidy
+ '[' 1 == 1 ']'
+ echo 'PR_CI_JOB set; skipping tidy'
+ SKIP_TIDY='--skip tidy'
+ ../x.py --stage 2 test --skip tidy --skip compiler --skip src
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2026-03-05/rustfmt-nightly-aarch64-unknown-linux-gnu.tar.xz
---
[RUSTC-TIMING] build_script_build test:false 0.148
[RUSTC-TIMING] cc test:false 0.706
   Compiling compiler_builtins v0.1.160 (/checkout/library/compiler-builtins/compiler-builtins)
[RUSTC-TIMING] build_script_build test:false 0.302
error[E0277]: the trait bound `T: [const] marker::Destruct` is not satisfied
   --> library/core/src/ops/try_trait.rs:445:21
    |
445 |     type Residual = NeverShortCircuitResidual;
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: required for `NeverShortCircuitResidual` to implement `[const] Residual<T>`
   --> library/core/src/ops/try_trait.rs:465:33
    |
465 | impl<T: [const] Destruct> const Residual<T> for NeverShortCircuitResidual {
    |         ----------------        ^^^^^^^^^^^     ^^^^^^^^^^^^^^^^^^^^^^^^^
    |         |
    |         unsatisfied trait bound introduced here
note: required by a bound in `Try::Residual`
   --> library/core/src/ops/try_trait.rs:160:20
    |
160 |     type Residual: [const] Residual<Self::Output>;
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Try::Residual`

For more information about this error, try `rustc --explain E0277`.
[RUSTC-TIMING] core test:false 21.864
error: could not compile `core` (lib) due to 1 previous error
Bootstrap failed while executing `--stage 2 test --skip tidy --skip compiler --skip src`
```

</details>

--- a/library/core/src/ops/try_trait.rs
+++ b/library/core/src/ops/try_trait.rs
@@ -157,7 +157,7 @@ pub const trait Try: [const] FromResidual {
     /// type: that type will have a "hole" in the correct place, and will maintain the
     /// "foo-ness" of the residual so other types need to opt-in to interconversion.
     #[unstable(feature = "try_trait_v2", issue = "84277", old_name = "try_trait")]
-    type Residual;
+    type Residual: Residual<Self::Output>;
 
     /// Constructs the type from its `Output` type.
     ///

--- a/src/tools/clippy/tests/ui/manual_try_fold.rs
+++ b/src/tools/clippy/tests/ui/manual_try_fold.rs
@@ -2,8 +2,9 @@
 #![allow(clippy::unnecessary_fold, unused)]
 #![warn(clippy::manual_try_fold)]
 #![feature(try_trait_v2)]
+#![feature(try_trait_v2_residual)]
 //@no-rustfix
-use std::ops::{ControlFlow, FromResidual, Try};
+use std::ops::{ControlFlow, FromResidual, Residual, Try};
 
 #[macro_use]
 extern crate proc_macros;
@@ -11,15 +12,21 @@ extern crate proc_macros;
 // Test custom `Try` with more than 1 argument
 struct NotOption(i32, i32);
 
+struct NotOptionResidual;
+
 impl<R> FromResidual<R> for NotOption {
     fn from_residual(_: R) -> Self {
         todo!()
     }
 }
 
+impl Residual<()> for NotOptionResidual {
+    type TryType = NotOption;
+}
+
 impl Try for NotOption {
     type Output = ();
-    type Residual = ();
+    type Residual = NotOptionResidual;
 
     fn from_output(_: Self::Output) -> Self {
         todo!()
@@ -34,15 +41,21 @@ impl Try for NotOption {
 #[derive(Default)]
 struct NotOptionButWorse(i32);
 
+struct NotOptionButWorseResidual;
+
 impl<R> FromResidual<R> for NotOptionButWorse {
     fn from_residual(_: R) -> Self {
         todo!()
     }
 }
 
+impl Residual<()> for NotOptionButWorseResidual {
+    type TryType = NotOptionButWorse;
+}
+
 impl Try for NotOptionButWorse {
     type Output = ();
-    type Residual = ();
+    type Residual = NotOptionButWorseResidual;
 
     fn from_output(_: Self::Output) -> Self {
         todo!()

--- a/src/tools/clippy/tests/ui/manual_try_fold.stderr
+++ b/src/tools/clippy/tests/ui/manual_try_fold.stderr
@@ -1,5 +1,5 @@
 error: usage of `Iterator::fold` on a type that implements `Try`
-  --> tests/ui/manual_try_fold.rs:59:10
+  --> tests/ui/manual_try_fold.rs:72:10
    |
 LL |         .fold(Some(0i32), |sum, i| sum?.checked_add(*i))
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use `try_fold` instead: `try_fold(0i32, |sum, i| ...)`
@@ -8,19 +8,19 @@ LL |         .fold(Some(0i32), |sum, i| sum?.checked_add(*i))
    = help: to override `-D warnings` add `#[allow(clippy::manual_try_fold)]`
 
 error: usage of `Iterator::fold` on a type that implements `Try`
-  --> tests/ui/manual_try_fold.rs:64:10
+  --> tests/ui/manual_try_fold.rs:77:10
    |
 LL |         .fold(NotOption(0i32, 0i32), |sum, i| NotOption(0i32, 0i32));
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use `try_fold` instead: `try_fold(..., |sum, i| ...)`
 
 error: usage of `Iterator::fold` on a type that implements `Try`
-  --> tests/ui/manual_try_fold.rs:68:10
+  --> tests/ui/manual_try_fold.rs:81:10
    |
 LL |         .fold(NotOptionButWorse(0i32), |sum, i| NotOptionButWorse(0i32));
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use `try_fold` instead: `try_fold(0i32, |sum, i| ...)`
 
 error: usage of `Iterator::fold` on a type that implements `Try`
-  --> tests/ui/manual_try_fold.rs:99:10
+  --> tests/ui/manual_try_fold.rs:112:10
    |
 LL |         .fold(Some(0i32), |sum, i| sum?.checked_add(*i))
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use `try_fold` instead: `try_fold(0i32, |sum, i| ...)`

--- a/tests/ui/consts/const-try.rs
+++ b/tests/ui/consts/const-try.rs
@@ -6,10 +6,11 @@
 
 #![crate_type = "lib"]
 #![feature(try_trait_v2)]
+#![feature(try_trait_v2_residual)]
 #![feature(const_trait_impl)]
 #![feature(const_try)]
 
-use std::ops::{ControlFlow, FromResidual, Try};
+use std::ops::{ControlFlow, FromResidual, Residual, Try};
 
 struct TryMe;
 struct Error;
@@ -29,6 +30,10 @@ impl const Try for TryMe {
     fn branch(self) -> ControlFlow<Self::Residual, Self::Output> {
         ControlFlow::Break(Error)
     }
+}
+
+impl Residual<()> for Error {
+    type TryType = TryMe;
 }
 
 const fn t() -> TryMe {

--- a/tests/ui/traits/const-traits/ice-126148-failed-to-normalize.rs
+++ b/tests/ui/traits/const-traits/ice-126148-failed-to-normalize.rs
@@ -1,6 +1,6 @@
 #![allow(incomplete_features)]
-#![feature(const_trait_impl, try_trait_v2, const_try)]
-use std::ops::{FromResidual, Try};
+#![feature(const_trait_impl, try_trait_v2, try_trait_v2_residual, const_try, const_try_residual)]
+use std::ops::{FromResidual, Residual, Try};
 
 struct TryMe;
 struct Error;
@@ -12,6 +12,10 @@ impl const Try for TryMe {
     //~^ ERROR not all trait items implemented
     type Output = ();
     type Residual = Error;
+}
+
+impl const Residual<()> for Error {
+    type TryType = TryMe;
 }
 
 const fn t() -> TryMe {

--- a/tests/ui/traits/const-traits/trait-default-body-stability.rs
+++ b/tests/ui/traits/const-traits/trait-default-body-stability.rs
@@ -5,11 +5,12 @@
 #![feature(const_trait_impl)]
 #![feature(const_t_try)]
 #![feature(const_try)]
+#![feature(const_try_residual)]
 #![feature(try_trait_v2)]
-
+#![feature(try_trait_v2_residual)]
 #![stable(feature = "foo", since = "1.0")]
 
-use std::ops::{ControlFlow, FromResidual, Try};
+use std::ops::{ControlFlow, FromResidual, Residual, Try};
 
 #[stable(feature = "foo", since = "1.0")]
 pub struct T;
@@ -27,6 +28,12 @@ impl const Try for T {
     fn branch(self) -> ControlFlow<T, T> {
         ControlFlow::Continue(self)
     }
+}
+
+#[stable(feature = "foo", since = "1.0")]
+#[rustc_const_unstable(feature = "const_t_try", issue = "none")]
+impl const Residual<T> for T {
+    type TryType = T;
 }
 
 #[stable(feature = "foo", since = "1.0")]

--- a/tests/ui/try-trait/try-operator-custom.rs
+++ b/tests/ui/try-trait/try-operator-custom.rs
@@ -1,12 +1,13 @@
 //@ run-pass
 
 #![feature(try_trait_v2)]
+#![feature(try_trait_v2_residual)]
 
-use std::ops::{ControlFlow, FromResidual, Try};
+use std::ops::{ControlFlow, FromResidual, Residual, Try};
 
 enum MyResult<T, U> {
     Awesome(T),
-    Terrible(U)
+    Terrible(U),
 }
 
 enum Never {}
@@ -27,7 +28,10 @@ impl<U, V> Try for MyResult<U, V> {
     }
 }
 
-impl<U, V, W> FromResidual<MyResult<Never, V>> for MyResult<U, W> where V: Into<W> {
+impl<U, V, W> FromResidual<MyResult<Never, V>> for MyResult<U, W>
+where
+    V: Into<W>,
+{
     fn from_residual(x: MyResult<Never, V>) -> Self {
         match x {
             MyResult::Terrible(e) => MyResult::Terrible(e.into()),
@@ -35,9 +39,16 @@ impl<U, V, W> FromResidual<MyResult<Never, V>> for MyResult<U, W> where V: Into<
     }
 }
 
+impl<U, V> Residual<U> for MyResult<Never, V> {
+    type TryType = MyResult<U, V>;
+}
+
 type ResultResidual<E> = Result<std::convert::Infallible, E>;
 
-impl<U, V, W> FromResidual<ResultResidual<V>> for MyResult<U, W> where V: Into<W> {
+impl<U, V, W> FromResidual<ResultResidual<V>> for MyResult<U, W>
+where
+    V: Into<W>,
+{
     fn from_residual(x: ResultResidual<V>) -> Self {
         match x {
             Err(e) => MyResult::Terrible(e.into()),
@@ -45,7 +56,10 @@ impl<U, V, W> FromResidual<ResultResidual<V>> for MyResult<U, W> where V: Into<W
     }
 }
 
-impl<U, V, W> FromResidual<MyResult<Never, V>> for Result<U, W> where V: Into<W> {
+impl<U, V, W> FromResidual<MyResult<Never, V>> for Result<U, W>
+where
+    V: Into<W>,
+{
     fn from_residual(x: MyResult<Never, V>) -> Self {
         match x {
             MyResult::Terrible(e) => Err(e.into()),

